src/main/java/org/openjdk/shenandoah/ShenandoahVisualizer.java

```diff
@@ -22,79 +23,354 @@
+        }
+
+        // Executors
+        ScheduledExecutorService service = Executors.newScheduledThreadPool(1);
+        final ScheduledFuture<?>[] f = {service.scheduleAtFixedRate(renderRunner,
+                0, renderRunner.isLive ? 100 : 1, TimeUnit.MILLISECONDS)};
 
-        Render render = new Render(data, frame);
 
         JPanel regionsPanel = new JPanel() {
             public void paint(Graphics g) {
-                render.renderRegions(g);
+                renderRunner.renderRegions(g);
             }
         };
 
         JPanel legendPanel = new JPanel() {
             @Override
             public void paint(Graphics g) {
-                Render.renderLegend(g);
+                renderRunner.renderLegend(g);
             }
         };
 
         JPanel statusPanel = new JPanel() {
             @Override
             public void paint(Graphics g) {
-                render.renderStats(g);
+                renderRunner.renderStats(g);
             }
         };
 
         JPanel graphPanel = new JPanel() {
             @Override
             public void paint(Graphics g) {
-                render.renderGraph(g);
+                renderRunner.renderGraph(g);
+            }
+        };
+
+        ActionListener realtimeModeButtonListener = new ActionListener() {
+            public void actionPerformed(ActionEvent ae) {
+                DataProvider data = new DataProvider(null);
+                renderRunner.loadLive(data);
+                toolbarPanel.setFileNameField("");
+                f[0] = changeScheduleInterval(100, service, f[0], renderRunner);
+            }
+        };
+        toolbarPanel.setRealtimeModeButtonListener(realtimeModeButtonListener);
+
+        ActionListener fileButtonListener = new ActionListener() {
+            public void actionPerformed(ActionEvent ae) {
+                JFileChooser fc = new JFileChooser();
+                int returnValue = fc.showOpenDialog(null);
+                int totalSnapshotSize = 0;
+                if (returnValue == JFileChooser.APPROVE_OPTION) {
+                    filePath[0] = fc.getSelectedFile().getAbsolutePath();
+                    try {
+                        DataLogProvider data = new DataLogProvider(filePath[0]);
+                        renderRunner.loadPlayback(data);
+                        totalSnapshotSize = data.getSnapshotsSize();
+                        toolbarPanel.setSize(totalSnapshotSize);
+                        toolbarPanel.setSnapshots(data.getSnapshots());
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                    renderRunner.playback.speed = 1.0;
+                    toolbarPanel.setSpeedValue(1.0);
+                    toolbarPanel.setFileNameField(filePath[0]);
+                    toolbarPanel.setLastActionField("File selected: " + filePath[0]);
+
+                    System.out.println("Selected file: " + filePath[0]);
```

Comment:
Is this debug output or is this useful? Not sure.

Reply:
It's debugging. I can have this use JUL?